mypy/meet.py

```diff
@@ -140,15 +149,28 @@ def visit_void(self, t: Void) -> Type:
             return ErrorType()
 
     def visit_none_type(self, t: NoneTyp) -> Type:
-        if not isinstance(self.s, Void) and not isinstance(self.s, ErrorType):
-            return t
+        if experiments.STRICT_OPTIONAL:
+            if isinstance(self.s, NoneTyp) or (isinstance(self.s, Instance) and
+                                               self.s.type.fullname() == 'builtins.object'):
+                return t
+            else:
+                return UninhabitedType()
         else:
-            return ErrorType()
+            if not isinstance(self.s, Void) and not isinstance(self.s, ErrorType):
+                return t
+            else:
+                return ErrorType()
+
+    def visit_uninhabited_type(self, t: UninhabitedType) -> Type:
+        return t
 
     def visit_deleted_type(self, t: DeletedType) -> Type:
         if not isinstance(self.s, Void) and not isinstance(self.s, ErrorType):
             if isinstance(self.s, NoneTyp):
-                return self.s
+                if experiments.STRICT_OPTIONAL:
+                    return t
+                else:
+                    return self.s
             else:
```

Comment:
Again I expect that the meet of `UninhabitedType` and a deleted type should be `UninhabitedType` (as `NoneTyp` has the same property when strict optional checking is disabled) but don't know exactly where it would matter.